File: src/main/java/com/github/shatteredsuite/scrolls/data/scroll/LocationBindingData.java
package com.github.shatteredsuite.scrolls.data.scroll;

import com.github.shatteredsuite.core.include.nbt.NBTCompound;
import com.github.shatteredsuite.scrolls.items.NBTUtils;
import com.github.shatteredsuite.scrolls.items.ScrollInstance;
import java.util.LinkedHashMap;
import java.util.Map;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.configuration.serialization.ConfigurationSerializable;
import org.bukkit.entity.Player;
import org.bukkit.event.player.PlayerTeleportEvent;
import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
import org.jetbrains.annotations.NotNull;

public class LocationBindingData extends BindingData implements ConfigurationSerializable {

    public final Location location;

    public LocationBindingData(Location location) {
        super("location");
        this.location = location;
    }

    @Override
    public ScrollInstance onInteract(ScrollInstance item, Player player) {
        PlayerTeleportEvent event = new PlayerTeleportEvent(player, player.getLocation(), location, TeleportCause.PLUGIN);
        Bukkit.getServer().getPluginManager().callEvent(event);
        if(!event.isCancelled()) {
            return item;
        }
        return new ScrollInstance(item.scrollType, item.isInfinite ? item.charges : item.charges-1, item.isInfinite, item.bindingData);
    }

    @Override
    public void applyBindingNBT(NBTCompound compound) {
        NBTUtils.addLocationNBT(compound, this.location);
    }

    @NotNull
    @Override
    public Map<String, Object> serialize() {
        Map<String, Object> map = new LinkedHashMap<>();
        return map;
    }
}

File: src/main/kotlin/com/github/shatteredsuite/scrolls/data/scroll/binding/LocationBindingData.kt
package com.github.shatteredsuite.scrolls.data.scroll.binding

import com.github.shatteredsuite.core.include.nbt.NBTCompound
import com.github.shatteredsuite.scrolls.ShatteredScrolls
import com.github.shatteredsuite.scrolls.data.ScrollCancelMode
import com.github.shatteredsuite.scrolls.items.NBTUtils
import com.github.shatteredsuite.scrolls.items.ScrollInstance
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.event.player.PlayerTeleportEvent
import java.util.*

class LocationBindingData(val location: Location) : BindingData("location", BindingDisplay("Teleportation Scroll", false, LinkedList(), false, 0)) {
    override fun onInteract(instance: ScrollInstance, player: Player): ScrollInstance {
        val event = PlayerTeleportEvent(player, player.location, location, PlayerTeleportEvent.TeleportCause.PLUGIN)
        Bukkit.getServer().pluginManager.callEvent(event)
        return if (!event.isCancelled) {
            if(location.world == null) {
                val inst = ShatteredScrolls.getInstance()
                return if (inst.config().cancelMode != ScrollCancelMode.UNBIND) {
                    inst.messenger.sendMessage(player, "unknown-world-unbind", true)
                    instance
                } else {
                    inst.messenger.sendMessage(player, "unknown-world-cancel", true)
                    ScrollInstance(instance.scrollType, instance.charges, instance.isInfinite, UnboundBindingData())
                }
            }
            player.teleport(location)
            ScrollInstance(instance.scrollType, instance.charges - 1, instance.isInfinite, instance.bindingData)
        }
        else instance
    }

    public override fun applyBindingNBT(compound: NBTCompound) {
        NBTUtils.addLocationNBT(compound, location)
    }

    override fun serialize(): Map<String?, Any?> {
        return mapOf("location" to this.location)
    }

    override fun parsePlaceholders(name: String): String {
        return name.replace("%x%", location.blockX.toString())
                .replace("%y%", location.blockY.toString())
                .replace("%z%", location.blockZ.toString())
                .replace("%yaw%", location.yaw.toString())
                .replace("%pitch%", location.pitch.toString())
                .replace("%world%", location.world?.name ?: "Unknown World")
    }
}